datarepo-clienttests/src/main/java/deploymentscripts/ModularHelmChart.java

```diff
@@ -0,0 +1,348 @@
+   * temporary files created in the current working directory.
+   *
+   * @param server the server configuration supplied by the test configuration
+   * @param app the application configuration supplied by the test configuration
+   */
+  public void deploy(ServerSpecification server, ApplicationSpecification app) throws Exception {
+    // store these on the instance to avoid passing them around to all the helper methods
+    serverSpecification = server;
+    applicationSpecification = app;
+
+    // get file handle to original/template API deployment Helm YAML file
+    File originalApiYamlFile =
+        FileUtils.createCopyOfFileFromURL(new URL(helmApiFilePath), "datarepo-api_ORIGINAL.yaml");
+
+    // modify the original/template YAML file and write the output to a new file
+    File modifiedApiYamlFile = FileUtils.createNewFile("datarepo-api_MODIFIED.yaml");
+    parseAndModifyApiYamlFile(originalApiYamlFile, modifiedApiYamlFile);
+
+    // delete the existing API deployment
+    // e.g. helm namespace delete mm-jade-datarepo-api --namespace mm
+    ArrayList<String> deleteCmdArgs = new ArrayList<>();
+    deleteCmdArgs.add("namespace");
+    deleteCmdArgs.add("delete");
+    deleteCmdArgs.add(serverSpecification.namespace + "-jade-datarepo-api");
+    deleteCmdArgs.add("--namespace");
+    deleteCmdArgs.add(serverSpecification.namespace);
+    Process helmDeleteProc = ProcessUtils.executeCommand("helm", deleteCmdArgs);
+    List<String> cmdOutputLines = ProcessUtils.waitForTerminateAndReadStdout(helmDeleteProc);
+    for (String cmdOutputLine : cmdOutputLines) {
+      System.out.println(cmdOutputLine);
+    }
+
+    // list the available deployments (for debugging)
+    // e.g. helm ls --namespace mm
+    ArrayList<String> listCmdArgs = new ArrayList<>();
+    listCmdArgs.add("ls");
+    listCmdArgs.add("--namespace");
+    listCmdArgs.add(serverSpecification.namespace);
+    Process helmListProc = ProcessUtils.executeCommand("helm", listCmdArgs);
+    cmdOutputLines = ProcessUtils.waitForTerminateAndReadStdout(helmListProc);
+    for (String cmdOutputLine : cmdOutputLines) {
+      System.out.println(cmdOutputLine);
+    }
+
+    // install/upgrade the API deployment using the modified YAML file we just generated
+    // e.g. helm namespace upgrade mm-jade-datarepo-api datarepo-helm/datarepo-api --install
+    // --namespace mm -f datarepo-api_MODIFIED.yaml
+    ArrayList<String> installCmdArgs = new ArrayList<>();
+    installCmdArgs.add("namespace");
+    installCmdArgs.add("upgrade");
+    installCmdArgs.add(serverSpecification.namespace + "-jade-datarepo-api");
+    installCmdArgs.add("datarepo-helm/datarepo-api");
+    installCmdArgs.add("--install");
+    installCmdArgs.add("--namespace");
+    installCmdArgs.add(serverSpecification.namespace);
+    installCmdArgs.add("-f");
+    installCmdArgs.add(modifiedApiYamlFile.getAbsolutePath());
+    Process helmUpgradeProc = ProcessUtils.executeCommand("helm", installCmdArgs);
```

Comment:
This is executing commands in the command line, right? Is there a java helm api like we're using for Kubernetes changes?

Reply:
Right. I didn't find an official or popular Java library. There is a GitHub project (microbean?) but it was last updated more than a year ago and I didn't want to rely on it. If you find or know of one though, then please let me know! It would be great to do this from Java.